cartographer/mapping/sparse_pose_graph.cc

```diff
@@ -38,24 +36,52 @@ proto::SparsePoseGraph::Constraint::Tag ToProto(
   LOG(FATAL) << "Unsupported tag.";
 }
 
+std::unordered_map<const Submaps*, int> IndexTrajectories(
+    const std::vector<const Submaps*>& trajectories) {
+  std::unordered_map<const Submaps*, int> result;
+  for (const auto& trajectory : trajectories) {
+    result.emplace(trajectory, result.size());
+  }
+  return result;
+}
+
 void GroupTrajectoryNodes(
     const std::vector<TrajectoryNode>& trajectory_nodes,
+    const std::unordered_map<const Submaps*, int>& trajectory_indices,
     std::vector<std::vector<TrajectoryNode>>* grouped_nodes,
     std::vector<std::pair<int, int>>* new_indices) {
   CHECK_NOTNULL(grouped_nodes)->clear();
   CHECK_NOTNULL(new_indices)->clear();
 
-  std::unordered_map<const Submaps*, int> trajectory_ids;
-  for (const auto node : trajectory_nodes) {
-    const auto* trajectory = node.constant_data->trajectory;
-    if (trajectory_ids.emplace(trajectory, grouped_nodes->size()).second) {
-      new_indices->emplace_back(grouped_nodes->size(), 0);
-      grouped_nodes->push_back({node});
-    } else {
-      const int id = trajectory_ids[trajectory];
-      new_indices->emplace_back(id, (*grouped_nodes)[id].size());
-      (*grouped_nodes)[id].push_back(node);
-    }
+  grouped_nodes->resize(trajectory_indices.size());
+
+  for (const auto& node : trajectory_nodes) {
+    auto index_iterator =
+        trajectory_indices.find(node.constant_data->trajectory);
+    CHECK(index_iterator != trajectory_indices.end());
+    const int index = index_iterator->second;
```

Comment:
`unordered_map::at()` throws if the key is not found, so simplify to:
```
const int id = trajectory_indices.at(node.constant_data->trajectory);
```

Similar in other places.